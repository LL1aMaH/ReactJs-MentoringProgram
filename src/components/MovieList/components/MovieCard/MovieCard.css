.card img {
  width: 100%;
  height: 100%;
}

.icon,
.buttons {
  display: none;
}

.card {
  position: relative;
}

.icon:hover {
  position: relative;
}

.card:hover .icon {
  position: absolute;
  right: 0;
  transform: rotate(90deg);
  font-size: 20px;
  background-color: var(--black-background-color);
  color: var(--primary-font-color);
  border-radius: 180%;
  width: 2rem;
  height: 2rem;
  text-align: center;
  cursor: default;
  margin: 0.5rem;
  display: block;
}

.icon:hover .buttons {
  display: block;
  transform: rotate(-90deg);
  right: -4.5rem;
  top: -1rem;
  border-radius: 4px;
  background-color: var(--primary-background-color);
  position: absolute;
  margin-top: 2rem;
}

.button {
  font-size: 15px;
  margin: 0.5rem 0;
  width: 100%;
  padding-right: 4rem;
  padding-left: 1rem;

  &:first-child {
    margin-top: 2rem;
  }

  &:hover {
    background-color: var(--primary-link-color);
  }
}
