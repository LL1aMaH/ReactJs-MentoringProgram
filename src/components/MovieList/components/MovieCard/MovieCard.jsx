import React, { memo, useCallback, useState } from 'react';

import { useRouter } from 'next/router'
// import { useHistory } from 'react-router-dom';

import poster from '../../../../../public/Movie_040311_3.jpg'

import { DeleteMovieModal } from '../../../Modals/DeleteMovieModal/DeleteMovieModal';
import { EditMovieModal } from '../../../Modals/EditMovieModal/EditMovieModal';

import { Button } from '../../../Button/Button';

import styles from '../../../../../styles/MovieCard.module.scss'

import { MovieInfo } from './components/MovieInfo/MovieInfo';


export const MovieCard = memo(function MovieCard({ film }){
  const { poster_path: posterURL, title, genres, release_date: date } = film;

  const router = useRouter()

  const [showEditModal, setShowEditModal] = useState(false);
  const [showDeleteModal, setShowDeleteModal] = useState(false);

  const editHandleClick = useCallback(() => {
    setShowEditModal((state) => !state);
  }, []);

  const deleteHandleClick = useCallback(() => {
    setShowDeleteModal((state) => !state);
  }, []);

  const onError = useCallback((e) => {
    e.currentTarget.src = '/src/assets/pictures/Movie_040311_3.jpg';
  }, []);

  const {id} = film
  const redirectToMovieDetails = useCallback(() => {
    router.push({ pathname: `/MovieDetailsPage/[id]`, query: { id } }, undefined)
  }, [
     film
  ]);


  return (
    <div className={styles.card}>
      {showEditModal && <EditMovieModal value={film} />}
      {showDeleteModal && <DeleteMovieModal id={id} />}

      <div className={styles.icon}>
        &#8230;
        <div className={styles.buttons}>
          <Button className={styles.button} onClick={editHandleClick} testId="Edit">
            Edit
          </Button>
          <Button className={styles.button} onClick={deleteHandleClick} testId="Delete">
            Delete
          </Button>
        </div>
      </div>
      <img
        src={posterURL || poster}
        alt="poster"
        onError={onError}
        onClick={redirectToMovieDetails}
      />
      <MovieInfo title={title} genres={genres} year={date.substr(0, 4)} />
    </div>
  );
});
