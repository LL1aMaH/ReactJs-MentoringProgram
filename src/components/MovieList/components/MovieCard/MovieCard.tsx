import React from 'react';

import poster from 'Assets/pictures/Movie_040311_3.jpg';

import { MovieInfo } from './components/MovieInfo';

import styles from './MovieCard.css';

export type MovieCardProps = {
  title: string;
  genres: Array<string>;
  year: number | string;
  posterURL?: string;
};

export const MovieCard = (props: MovieCardProps): JSX.Element => {
  const { title, genres, year, posterURL } = props;
  return (
    <div className={styles.card}>
      <img src={posterURL || poster} alt="poster" />
      <MovieInfo title={title} genres={genres} year={year} />
    </div>
  );
};
