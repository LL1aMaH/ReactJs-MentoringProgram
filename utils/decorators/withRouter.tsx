import React from 'react';

import { MemoryRouter } from 'react-router-dom';

export type TComponentFunc = () => JSX.Element;
export type TDecoratorLocation = {
  pathname?: string;
  search?: string;
};

const withRouter = (
  componentFunc: TComponentFunc,
  location: TDecoratorLocation = { pathname: '/' },
): JSX.Element => {
  return <MemoryRouter initialEntries={[location]}>{componentFunc()}</MemoryRouter>;
};

export default withRouter;
