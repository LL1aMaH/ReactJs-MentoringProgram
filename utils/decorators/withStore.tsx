import React from 'react';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import originalStore from './configureStore';

type TComponentFunc = () => JSX.Element;

const withStore = (componentFunc: TComponentFunc, initialState = {}): JSX.Element => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares);
  const store = mockStore(initialState);
  return <Provider store={store}>{componentFunc()}</Provider>;
};

export const withOriginalStore = (componentFunc: TComponentFunc): JSX.Element => {
  return <Provider store={originalStore}>{componentFunc()}</Provider>;
};

export default withStore;
