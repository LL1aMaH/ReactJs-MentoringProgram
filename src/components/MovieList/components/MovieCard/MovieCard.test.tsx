import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import getFormDecorator from '../../../../../utils/decorators/withForm';
import withRouter from '../../../../../utils/decorators/withRouter';
import withStore from '../../../../../utils/decorators/withStore';

import { MovieCard } from './MovieCard';

describe('MovieCard', () => {
  const withForm = getFormDecorator({ initialValues: { test: 'test' } });
  it('should renders correctly', () => {
    const component = render(
      withRouter(() => (
        <MovieCard
          film={{
            id: 447365,
            title: 'Guardians of the Galaxy Vol. 3',
            release_date: '2020-05-01',
            overview: "The third film based on Marvel's Guardians of the Galaxy.",
            runtime: 100,
            genres: [],
            tagline: '',
            vote_average: 10,
            vote_count: 9,
            poster_path: 'https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg',
            budget: 0,
            revenue: 0,
          }}
        />
      )),
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should renders edit modal correctly', () => {
    const component = render(
      withStore(
        () =>
          withForm(() =>
            withRouter(() => (
              <MovieCard
                film={{
                  id: 447365,
                  title: 'Guardians of the Galaxy Vol. 3',
                  release_date: '2020-05-01',
                  overview: "The third film based on Marvel's Guardians of the Galaxy.",
                  runtime: 100,
                  genres: [],
                  tagline: '',
                  vote_average: 10,
                  vote_count: 9,
                  poster_path: 'https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg',
                  budget: 0,
                  revenue: 0,
                }}
              />
            )),
          ),
        {
          genre: 'ALL',
          sort: 'RELEASE DATE',
        },
      ),
    );

    const button = component.getByTestId('Edit');
    fireEvent.click(button);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should renders delete modal correctly', () => {
    const component = render(
      withStore(
        () =>
          withForm(() =>
            withRouter(() => (
              <MovieCard
                film={{
                  id: 447365,
                  title: 'Guardians of the Galaxy Vol. 3',
                  release_date: '2020-05-01',
                  overview: "The third film based on Marvel's Guardians of the Galaxy.",
                  runtime: 100,
                  genres: [],
                  tagline: '',
                  vote_average: 10,
                  vote_count: 9,
                  poster_path: '',
                  budget: 0,
                  revenue: 0,
                }}
              />
            )),
          ),
        {
          genre: 'ALL',
          sort: 'RELEASE DATE',
        },
      ),
    );

    const button = component.getByTestId('Delete');
    fireEvent.click(button);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should renders correctly with onError event', () => {
    const { getByAltText } = render(
      withRouter(() => (
        <MovieCard
          film={{
            id: 447365,
            title: 'Guardians of the Galaxy Vol. 3',
            release_date: '2020-05-01',
            overview: "The third film based on Marvel's Guardians of the Galaxy.",
            runtime: 100,
            genres: [],
            tagline: '',
            vote_average: 10,
            vote_count: 9,
            poster_path: 'https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg',
            budget: 0,
            revenue: 0,
          }}
        />
      )),
    );
    const image: any = getByAltText('poster');
    fireEvent.error(image);
    expect(image.src).toBe('http://localhost/src/assets/pictures/Movie_040311_3.jpg');
  });

  it('should redirect correctly', () => {
    const component = render(
      withStore(
        () =>
          withForm(() =>
            withRouter(() => (
              <MovieCard
                film={{
                  id: 447365,
                  title: 'Guardians of the Galaxy Vol. 3',
                  release_date: '2020-05-01',
                  overview: "The third film based on Marvel's Guardians of the Galaxy.",
                  runtime: 100,
                  genres: [],
                  tagline: '',
                  vote_average: 10,
                  vote_count: 9,
                  poster_path: '',
                  budget: 0,
                  revenue: 0,
                }}
              />
            )),
          ),
        {
          genre: 'ALL',
          sort: 'RELEASE DATE',
        },
      ),
    );

    const image = component.getByAltText('poster');
    fireEvent.click(image);

    expect(component.asFragment()).toMatchSnapshot();
  });
});
