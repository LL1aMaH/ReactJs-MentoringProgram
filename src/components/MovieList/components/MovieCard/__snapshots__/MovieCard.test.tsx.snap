// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should redirect correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="icon"
    >
      …
      <div
        class="buttons"
      >
        <button
          class="button button"
          data-testid="Edit"
          tabindex="0"
        >
          Edit
        </button>
        <button
          class="button button"
          data-testid="Delete"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <img
      alt="poster"
      src="[object Object]"
    />
    <div
      class="card"
    >
      <div
        class="info"
      >
        <p
          class="title"
        >
          Guardians of the Galaxy Vol. 3
        </p>
        <p
          class="genres"
        />
      </div>
      <p
        class="year"
      >
        2020
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieCard should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="icon"
    >
      …
      <div
        class="buttons"
      >
        <button
          class="button button"
          data-testid="Edit"
          tabindex="0"
        >
          Edit
        </button>
        <button
          class="button button"
          data-testid="Delete"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <img
      alt="poster"
      src="https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg"
    />
    <div
      class="card"
    >
      <div
        class="info"
      >
        <p
          class="title"
        >
          Guardians of the Galaxy Vol. 3
        </p>
        <p
          class="genres"
        />
      </div>
      <p
        class="year"
      >
        2020
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieCard should renders delete modal correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="icon"
    >
      …
      <div
        class="buttons"
      >
        <button
          class="button button"
          data-testid="Edit"
          tabindex="0"
        >
          Edit
        </button>
        <button
          class="button button"
          data-testid="Delete"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <img
      alt="poster"
      src="[object Object]"
    />
    <div
      class="card"
    >
      <div
        class="info"
      >
        <p
          class="title"
        >
          Guardians of the Galaxy Vol. 3
        </p>
        <p
          class="genres"
        />
      </div>
      <p
        class="year"
      >
        2020
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieCard should renders edit modal correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="icon"
    >
      …
      <div
        class="buttons"
      >
        <button
          class="button button"
          data-testid="Edit"
          tabindex="0"
        >
          Edit
        </button>
        <button
          class="button button"
          data-testid="Delete"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <img
      alt="poster"
      src="https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg"
    />
    <div
      class="card"
    >
      <div
        class="info"
      >
        <p
          class="title"
        >
          Guardians of the Galaxy Vol. 3
        </p>
        <p
          class="genres"
        />
      </div>
      <p
        class="year"
      >
        2020
      </p>
    </div>
  </div>
</DocumentFragment>
`;
